pkg/karmadactl/apply.go

```diff
@@ -0,0 +1,190 @@
+package karmadactl
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/resource"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubectl/pkg/cmd/apply"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"k8s.io/kubectl/pkg/util/templates"
+
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/karmadactl/options"
+	"github.com/karmada-io/karmada/pkg/util/names"
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+// CommandApplyOptions contains the input to the apply command.
+type CommandApplyOptions struct {
+	// global flags
+	options.GlobalCommandOptions
+}
+
+var (
+	applyLong = templates.LongDesc(i18n.T(`
+		Apply a configuration to a resource by file name or stdin.
+		The resource name must be specified. This resource will be created if it doesn't exist yet.
+		To use 'apply', always create the resource initially with either 'apply' or 'create --save-config'.
+
+		JSON and YAML formats are accepted.
+
+		Alpha Disclaimer: the --prune functionality is not yet complete. Do not use unless you are aware of what the current state is. See https://issues.k8s.io/34274.`))
+
+	applyExample = templates.Examples(i18n.T(`
+		# Apply the configuration in pod.json to a pod
+		%[1]s apply -f ./pod.json
+
+		# Apply resources from a directory containing kustomization.yaml - e.g. dir/kustomization.yaml
+		%[1]s apply -k dir/
+
+		# Apply the JSON passed into stdin to a pod
+		cat pod.json | %[1]s apply -f -
+
+		# Note: --prune is still in Alpha
+		# Apply the configuration in manifest.yaml that matches label app=nginx and delete all other resources that are not in the file and match label app=nginx
+		%[1]s apply --prune -f manifest.yaml -l app=nginx
+
+		# Apply the configuration in manifest.yaml and delete all the other config maps that are not in the file
+		%[1]s apply --prune -f manifest.yaml --all --prune-whitelist=core/v1/ConfigMap`))
+)
+
+// NewCmdApply creates the `apply` command
+// TODO(@carlory): use kcmdutil.factory as an input argument instead of the karmada config,
+// in order to make it act like the kubectl apply command.
```

Comment:
We seemly don't need this `TODO`, because we need to build new `kcmdutil.factory` in some subcmd, we don't need to make it act like the kubectl apply command.